Add XNNPACK Clamp.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36202 Enable torch_speed_benchmark to accept different memory formats.
* **#35896 Add XNNPACK Clamp.**
* #35791 Enable stateless XNNPACK linear.
* #35790 Enable stateless XNNPACK convolutions.
* #36032 Mobile CPU allocator.

Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

Commit:
Add XNNPACK Clamp.

[ghstack-poisoned]

Commit:
Update on "[WIP] Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "[WIP] Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "[WIP] Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "[WIP] Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "[WIP] Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -32,14 +33,47 @@ DEFINE_DISPATCH(leaky_relu_stub);
 DEFINE_DISPATCH(leaky_relu_backward_stub);
 
 Tensor hardtanh(const Tensor& self, Scalar min, Scalar max) {
+// #if defined(C10_MOBILE)
```
Given that this are runtime ops, and not the ones which are folded/fused into conv/linear, you might want to just do this at clamp? https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Activation.cpp#L35. My impression is that in tracing, it will trace `clamp` and you may not see hardtanh at all. Same for relu's `threshold`, I presume.

AshkanAliabadi replied:
I wanted to do that but I felt that would be even more intrusive.  I might be wrong.

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

dreiss reviewed (changes requested):
This seems like a lot of code for an op that we expect to usually be fused or memory-bound.  Can you update the commit message to give some more context on why this is needed and what the benefits are?  https://fb.workplace.com/groups/ToffeeInternal/permalink/472875023290983/ has some good guidelines.

AshkanAliabadi commented:
Will update the commit message.  As for the why, performance is like Physics: observations are everything, and explanations are only valid as long as they fit the observation, and only valuable so long as they give us some predictive power but other than that have no intrinsic value.  It is true that ReLUs should be memory bound, but there is a 9 millisecond difference on FBNet between XNNPACK's implementation and PyTorch's.  Why this is happening must have to do the with quality of implementation, but this difference is there nonetheless.  The question is whether we want to leave that much performance on the table to keep the code simpler. :)

It is also true that ReLUs should be fused, making this code path redundant, but not all use-cases are currently fuse-able like ReLUs that follow max pools, and without this code path those use cases will go through the slower route.

Also not all users seem to be running optimization passes on mobile mostly because we should do a better job communicating how they should do that.  In this regard, I have added this code path for the same reason I added stateless convolutions and linears: to avoid leaving the user with shockingly bad performance even if there are more optimal ways of achieving the same goal.

kimishpatel commented:
@dreiss, I agree with @AshkanAliabadi here. Yes these ops should be memory bound but inefficiencies can come from the way op is implemented. Maybe there are some redundant things being done. Calls to contiguous etc. In floating point requant suff that I was working on, I was surprised to find that doing requant via floating point path made it much slower. Part of this might be because we made it compute bound (?) because the latency to do FP multiply-add is double that of int counterpart. My guesses.
On the part of fusing clamps with previous op: I was thinking for all ops defined in `_xnnpack` namespace we should be able to fuse them. I was thinking of figuring out a way to write a generic JIT pass that can fuse such instances without having to write separate passes for each such op.

dreiss commented:
I think we need to do more homework before pushing this through.  PyTorch Mobile is meant to be a unified framework with non-mobile PyTorch, and the more special cases we introduce for mobile, the more divergence and confusion we will have.  We don't want PT to become a mass of special cases.  I think we all agree that we will need them for Conv and Linear, but for simple ops like this, I think we need to at least get an explanation for why the existing implementation can't be fixed before we create a special case.

kimishpatel commented:
> I think we need to do more homework before pushing this through. PyTorch Mobile is meant to be a unified framework with non-mobile PyTorch, and the more special cases we introduce for mobile, the more divergence and confusion we will have. We don't want PT to become a mass of special cases. I think we all agree that we will need them for Conv and Linear, but for simple ops like this, I think we need to at least get an explanation for why the existing implementation can't be fixed before we create a special case.

So quick glance here:https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/Activation.cpp#L85, and here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/vec256_base.h#L126,. Some of the unary and binary ops are implemented via tensor iterator which eventually use this Vec256 class to vectorize ops. Unfortunately they are only available for x86, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/vec256_float.h#L40, else we fallback to default implementation pointed above. I dont think compiler is able to vectorize those. Again we wont know for sure without looking at the asm (we can confirm this) but this is my guess. I mentioned this recently that maybe we should have neon backend for Vec256, and Jianyu had brought this up before too. Probably thats the reason why those are slow.

dreiss commented:
> I mentioned this recently that maybe we should have neon backend for Vec256

Sounds good to me.  @jamesr66a , this was part of your plan to reduce the divergence between QNNPACK and fbgemm also, right?

AshkanAliabadi commented:
@dreiss I would prefer if we move forward with this change and then remove it once that backend is ready in case that change will take more than a week or two at most to merge.  My goal is to get to a state of good performance as soon as possible and not make achieving good performance on mobile contingent on long term proejcts.  If and when that backend is ready, I would be more than happy to remove this.

jamesr66a commented:
> > I mentioned this recently that maybe we should have neon backend for Vec256
> 
> Sounds good to me. @jamesr66a , this was part of your plan to reduce the divergence between QNNPACK and fbgemm also, right?

Yes, but the *NNPACK strike force made impressive progress since then :) I think it would still make sense to have a NEON backend, though

Commit:
Update on "Add XNNPACK Clamp."


Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight
prepacking.  If we have done our job properly, JIT must have caught and
replaced nn.Conv2d's on mobile with their corresponding prepacked
versions, hence enabling said optimal usage.  Still, if we somehow end
up in at::native::_convolutions for whatever reason, it is still more
efficient to sub-optimally perform a stateless convolution on
channels-last tensors compared to the alternative of converting
channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW
tensors though, would be a toss up between NNPACK and XNNPACK, yet
in favor of reducing binary size as a result of removing dependence on
NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight
prepacking.  If we have done our job properly, JIT must have caught and
replaced nn.Conv2d's on mobile with their corresponding prepacked
versions, hence enabling said optimal usage.  Still, if we somehow end
up in at::native::_convolutions for whatever reason, it is still more
efficient to sub-optimally perform a stateless convolution on
channels-last tensors compared to the alternative of converting
channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW
tensors though, would be a toss up between NNPACK and XNNPACK, yet
in favor of reducing binary size as a result of removing dependence on
NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight
prepacking.  If we have done our job properly, JIT must have caught and
replaced nn.Conv2d's on mobile with their corresponding prepacked
versions, hence enabling said optimal usage.  Still, if we somehow end
up in at::native::_convolutions for whatever reason, it is still more
efficient to sub-optimally perform a stateless convolution on
channels-last tensors compared to the alternative of converting
channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW
tensors though, would be a toss up between NNPACK and XNNPACK, yet
in favor of reducing binary size as a result of removing dependence on
NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/xnnpack/Engine.h:
```diff
@@ -6,6 +6,90 @@ namespace at {
+    const Tensor& input,
+    float output_min,
+    float output_max);
+
```
The following two functions are a handful, but I am introducing complexity here for the benefit of minimal modifications to the call sites.

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

Commit:
Update on "Add XNNPACK Clamp."


Optimal usage of channels-last convolutions would require weight prepacking.  If we have done our job properly, JIT must have caught and replaced nn.Conv2d's on mobile with their corresponding prepacked versions, hence enabling said optimal usage.  Still, if we somehow end up in at::native::_convolutions for whatever reason, it is still more efficient to sub-optimally perform a stateless convolution on channels-last tensors compared to the alternative of converting channels-last (NHWC) to contigous (NCHW) and going through NNPACK.  NCHW tensors though, would be a toss up between NNPACK and XNNPACK, yet in favor of reducing binary size as a result of removing dependence on NNPACK, it is still better to use XNNPACK on mobile.

Differential Revision: [D20821865](https://our.internmc.facebook.com/intern/diff/D20821865)

[ghstack-poisoned]

AshkanAliabadi commented:
I removed all integration with call sites.  Is this a good compromise?  At least if push comes to shove and we really need a fast path for a use-case this is readily available.
